impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java

```diff
@@ -354,6 +370,29 @@ public synchronized void close() throws Exception {
         }
     }
 
+    /**
+     * Workaround for <a href="https://issues.apache.org/jira/browse/KAFKA-14287">KAFKA-14287</a>.
+     * with kraft, if we don't shut down the controller last, we sometimes see a hang.
+     */
+    private void roleOrderedShutdown(Map<Integer, Server> servers, boolean await) {
```

Comment:
I take that as yes to 1. but we don't see it due to the hard kill.

Do you see what I'm worrying about in 2?

Reply:
1. I expect so.
2. Yes, I do.  At a minimum there might be an opportunity for a pull-up type refactor to factor out the duplication around the server/broker maps.   Like you say, it might be better modelled differently.    I don't feel the motivation to make this change at the moment tbh.

Reply:
Re 2. As long as we are on the same page. I'll put trying to tease things about on my todo list but I want to try and make some progress on shared state first.